### ISSUE TITLE
Shutdown hangs unmounting /var/lib/nfs/rpc_pipefs

### ISSUE THREAD

**Issue by iandall** [systemd/systemd#28410]
### systemd version the issue has been seen with

254

### Linux kernel version used

6.3.12-200.fc38.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

Clean shutdown.

### Unexpected behaviour you saw

Sometimes (but not always) hangs attempting to unmount /var/lib/nfs/rpc_pipefs. Timeout is infinite.

### Steps to reproduce the problem

Presumably only happens on a system using NFS. This system has NFS root and /var is also NFS mounted. Simply run `shutdown -h now`

**Comment by iandall** [comment]
In earlier version (253.5-1.fc38) there was no hang, but the unmount of /var/lib/nfs/rpc_pipefs failed with status 32 (EPIPE) and subsequently the unmount of /var failed with status 16 (EBUSY). At least they failed quickly!

I am confused, because there is no var-lib-nfs-rpc_pipefs.mount unit file I (or systemctl) can find. There is a rpc-pipefs-generator (from nfs-utils), but there is no corresponding unit file in /run. Does the generator ever run? Nonetheless you can do `systemctl status  var-lib-nfs-rpc_pipefs.mount` or `var-lib-nfs-rpc_pipefs.mount`, so I am assuming the rpc_pipefs is a known special filesystem.

Would it help to put rpc_pipefs under /run? At least then /var could be unmounted.

**Comment by iandall** [comment]
OK, rpc_pipefs is mounted from a shell script in initrd and moved to sysroot before switch-root, so this is not done with systemd unit files (either static or generated).

**Comment by bluca** [comment]
Please indicate the distribution you are using and provide debug level logs

**Comment by sciolistten** [comment]
Just a quick addition to this issue, as we are seeing similar problems:

`[root@ourhost logs]# uname -a
Linux vm-bmsprdpdb-01.internal.yale.edu 3.10.0-1160.92.1.el7.x86_64 #1 SMP Thu May 18 11:23:40 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
[root@ourhost logs]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.9 (Maipo)`

The rpc_pipefs.target file (/usr/lib/systemd/system/rpc_pipefs.target) is provided by the nfs-utils package on this system.
`[root@ourhost logs]# rpm -qf /usr/lib/systemd/system/rpc_pipefs.target
nfs-utils-1.3.0-0.68.el7.2.x86_64`

I have seen reference to an issue in the nfs at https://patchwork.kernel.org/project/linux-nfs/patch/20190125083302.3346-1-yongcheng.yang@gmail.com/

I haven't implemented the suggested fix for that yet, but was wondering why the "BindsTo" behavior would differ from the "Requires" behavior.

**Comment by sciolistten** [comment]
doh! reading that fine manual answered my question:

> it also declares that this unit is stopped when any of the units listed suddenly
           disappears. Units can suddenly, unexpectedly disappear if a service terminates on its own choice, a device is unplugged or a mount point unmounted without involvement of systemd.

**Comment by iandall** [comment]
Distribution is: Fedora release 38 (Thirty Eight)

The log is terminated uncleanly, but here it is: [unmount-rpc_pipefs.log](https://github.com/systemd/systemd/files/12189601/unmount-rpc_pipefs.log)

**Comment by poettering** [comment]
So this is interesting:

```
[157900.361936] systemd[1]: var-lib-nfs-rpc_pipefs.mount: Child 137945 belongs to var-lib-nfs-rpc_pipefs.mount.
[157900.361994] systemd[1]: var-lib-nfs-rpc_pipefs.mount: Mount process exited, code=exited, status=0/SUCCESS (success)
[157900.362067] systemd[1]: var-lib-nfs-rpc_pipefs.mount: Mount still present, trying again.
[157900.362133] systemd[1]: var-lib-nfs-rpc_pipefs.mount: Path '/var/lib/nfs/rpc_pipefs' is not a mount point anymore, waiting for mount table refresh.
[157900.362188] systemd[1]: var-lib-nfs-rpc_pipefs.mount: Changed unmounting -> unmounting-catchup
```
So it appears the umount succeeds (i.e. returns 0). But the mount actually is still there. Which we detect and then try to unmount again. Except that once we are about to run umount again its gone after all, and things hang. This suggests two bugs:

1. for some reason umount for rpc_pipefs is async? i.e. it returns before the mount is actually gone? which is super weird? smells like an nfs/mount bug, dunno.
2. systemd gets confused, and expects a mount table refresh that never comes... this part is a systemd bug.

**Comment by yuwata** [comment]
Setting 'regression' label, as this is caused by 1483892a421ca34bc841a8e8b1f385744c0407ed.

**Comment by yuwata** [comment]
Fix is waiting in #28751.
